src/transformers/data/data_collator.py

```diff
@@ -313,3 +314,136 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor,
             ) & masked_indices[i]
 
         return inputs, perm_mask, target_mapping, labels
+
+
+@dataclass
+class DataCollatorForNSP:
+    """
+    Data collator used for language modeling.
+    - collates batches of tensors, honoring their tokenizer's pad_token
+    - preprocesses batches for masked language modeling
+    """
+
+    tokenizer: PreTrainedTokenizer
+    mlm: bool = True
+    nsp_probability: float = 0.5
+    block_size: int = 128
+    short_seq_prob: float = 0.1
+
+    def __call__(self, examples: List[List[List[int]]]) -> Dict[str, torch.Tensor]:
+        input_ids = []
+        segment_ids = []
+        labels = []
+
+        for i, doc in enumerate(examples):
+            input_id, segment_id, label = self.create_examples_from_document(doc, i, examples)
+            input_ids.extend(input_id)
+            segment_ids.extend(segment_id)
+            labels.extend(label)
+
+        return {
+            "input_ids": self._tensorize_batch(input_ids),
+            "token_type_ids": self._tensorize_batch(segment_ids),
+            "next_sentence_label": torch.tensor(labels),
+        }
+
+    def _tensorize_batch(self, examples: List[torch.Tensor]) -> torch.Tensor:
+        length_of_first = examples[0].size(0)
+        are_tensors_same_length = all(x.size(0) == length_of_first for x in examples)
+        if are_tensors_same_length:
+            return torch.stack(examples, dim=0)
+        else:
+            if self.tokenizer._pad_token is None:
+                raise ValueError(
+                    "You are attempting to pad samples but the tokenizer you are using"
+                    f" ({self.tokenizer.__class__.__name__}) does not have one."
+                )
+            return pad_sequence(examples, batch_first=True, padding_value=self.tokenizer.pad_token_id)
+
+    def create_examples_from_document(
+        self, document: List[List[int]], doc_index: int, examples: List[List[List[int]]]
+    ):
+        """Creates examples for a single document."""
+
+        # Account for [CLS], [SEP], [SEP]
+        max_num_tokens = self.block_size - 3
```

Comment:
Shouldn't this be `block_size - tokenizer.num_special_tokens_to_add(pair=False)` to work with any tokenizer?